tests/test_e2e.py

```diff
@@ -58,56 +57,35 @@ def test_e2e_plan(cli_runner: CliRunner, data_dir: Path, rules_dir: Path) -> Non
     assert result.exit_code == 0
 
 
-@freeze_time("2023-05-12 12:12:53")
-@pytest.mark.timeout(5)
 def test_e2e_gui(
-    cli_runner: CliRunner,
     unused_tcp_port: int,
     data_dir: Path,
     tmp_path: Path,
-    thread_executor: ThreadPoolExecutor,
-    mocker: MockerFixture,
 ) -> None:
-    def client_select_directory(port: int) -> httpx.Response:
-        # It's probably overkill to start an event loop,
-        # but wait_for_port provides exactly what's needed here.
-        # Use a timeout so a failing test won't be held up by this thread.
-        asyncio.run(asyncio.wait_for(wait_for_port(port), timeout=2))
-        return httpx.post(
-            f"http://127.0.0.1:{port}/redact/",
-            data={
-                "input_directory": str(data_dir / "input" / "tiff"),
-                "output_directory": str(tmp_path),
-            },
-        )
-
-    webbrowser_open_mock = mocker.patch("webbrowser.open")
-    # Ideally, we'd be able to use a running event loop to schedule the client
-    # requests. However, the CLI starts its own event loop, then blocks until
-    # completion. We could hook into or patch internal application events to
-    # get a callback with an already-running event loop, but that provides less
-    # isolation than just running the client in an independent thread.
-    # Note, Click's CliRunner expects to run in a single-threaded environment,
-    # but the new thread won't use any of the patched stdout, etc. streams and
-    # a ProcessPoolExecutor is slower and requires more delicate pickling. so
-    # we'll break that expection here.
-    client_future = thread_executor.submit(client_select_directory, unused_tcp_port)
-
-    cli_result = cli_runner.invoke(main.imagedephi, ["gui", "--port", str(unused_tcp_port)])
-    # If an HTTP request is successfully sent, either the normal response from this endpoint or a
-    # 500 error should cause the CLI to halt. However, if halting fails, then this test is
-    # particularly susceptible to exceeding the timeout, which on Windows will crash pytest (due
-    # to the behavior of pytest-timeout).
-
-    assert cli_result.exit_code == 0
-    webbrowser_open_mock.assert_called_once()
-    output_file = tmp_path / "Redacted_2023-05-12_12-12-53" / "study_slide_1.tif"
+
+    port = unused_tcp_port
+    subprocess.Popen(
+        [sys.executable, "-m", "imagedephi", "gui", "--port", str(port)],
+
+    )
+    sleep(2)
```

Comment:
So will this test always take 4 seconds at minimum, given the sleeps?

Is there a way to make it closer to the previous behavior, where it polled for an open socket or the termination of the CLI (or maybe just the existence of a output file)?

If test cases contain `sleep`, they're guaranteed to be much slower than every other test case for local development (causing overall test suite times to go up, sometimes by an order of magnitude). Sometimes in a CI environment with very limited resources, the hardcoded sleep duration is not enough, and the test will fail.

If there's no way around `sleep`, there's an option to use [custom Pytest markers)(https://docs.pytest.org/en/7.1.x/example/markers.html) to create groups like `slow` or `flaky`, which can be excluded for some test runs. Alternatively, you can use [OS-specific test cases](https://github.com/DigitalSlideArchive/ImageDePHI/blob/28dad54eb8fbf9f28db8873fd004baf91454a37e/tests/test_utils_os.py#L18) to include or exclude tests based on the OS, or even an OS-specific parameterized fixture which provides setup and teardown in different ways for a single consolidated test case.

Reply:
fixed in c95f49ee9c4caa7ae50536b73b038fcb266dcc81